core/src/family.cc

```diff
@@ -70,19 +74,89 @@ const std::map<std::string, std::string> Family<T>::GetConstantLabels() const {
 
 template <typename T>
 std::vector<MetricFamily> Family<T>::Collect() {
-  std::lock_guard<std::mutex> lock{mutex_};
   auto family = MetricFamily{};
-  family.name = name_;
-  family.help = help_;
-  family.type = T::metric_type;
-  for (const auto& m : metrics_) {
-    family.metric.push_back(std::move(CollectMetric(m.first, m.second.get())));
+  std::vector<std::shared_ptr<T>> to_be_removed;
+  {
+    std::lock_guard<std::mutex> lock{mutex_};
+    family.name = name_;
+    family.help = help_;
+    family.type = T::metric_type;
+    for (const auto& m : metrics_) {
+      if (!m.second->Expired()) {
+        family.metric.push_back(std::move(CollectMetric(m.first, m.second)));
+      } else if (retention_behavior_ == RetentionBehavior::Remove) {
+        to_be_removed.push_back(m.second);
```

Comment:
I get that we don't want to hold the lock for a long time but if you're using unordered maps for the metrics and labels, then the removal should be constant time in the average case. If removal from the unordered maps approach linear time then I think a better hash function needs to be chosen. Hence, do we need to store the metrics to be removed and then remove them later?

Reply:
That could be a solution. The Remove method does a few things, so I think I was just playing it safe. Additionally, the Remove method also has a lock, so that would either need to be removed (and only called when the data is locked, which is problematic) or change it to a scoped lock, which would need some thought.

I like the idea of calling it in place but I'm leaning towards leaving it since I like to play it safe with locks.